src/Cryptol/Testing/Random.hs

```diff
@@ -167,55 +180,236 @@ randomSize k n g
+randomIntMod :: (Backend sym, RandomGen g) => sym -> Integer -> Gen g sym
+randomIntMod sym modulus _ g =
+  let (i, g') = randomR (0, modulus-1) g
+  in (VInteger <$> integerLit sym i, g')
 
 -- | Generate a random word of the given length (i.e., a value of type @[w]@)
 -- The size parameter is assumed to vary between 1 and 100, and we use
 -- it to generate smaller numbers first.
-randomWord :: (BitWord b w i, RandomGen g) => Integer -> Gen g b w i
-randomWord w _sz g =
+randomWord :: (Backend sym, RandomGen g) => sym -> Integer -> Gen g sym
+randomWord sym w _sz g =
    let (val, g1) = randomR (0,2^w-1) g
-   in (VWord w (ready (WordVal (wordLit w val))), g1)
+   in (return $ VWord w (WordVal <$> wordLit sym w val), g1)
 
 -- | Generate a random infinite stream value.
-randomStream :: RandomGen g => Gen g b w i -> Gen g b w i
+randomStream :: (Backend sym, RandomGen g) => Gen g sym -> Gen g sym
 randomStream mkElem sz g =
   let (g1,g2) = split g
-  in (VStream $ IndexSeqMap $ genericIndex (map ready (unfoldr (Just . mkElem sz) g1)), g2)
+  in (pure $ VStream $ IndexSeqMap $ genericIndex (unfoldr (Just . mkElem sz) g1), g2)
 
 {- | Generate a random sequence.  This should be used for sequences
 other than bits.  For sequences of bits use "randomWord". -}
-randomSequence :: RandomGen g => Integer -> Gen g b w i -> Gen g b w i
+randomSequence :: (Backend sym, RandomGen g) => Integer -> Gen g sym -> Gen g sym
 randomSequence w mkElem sz g0 = do
   let (g1,g2) = split g0
   let f g = let (x,g') = mkElem sz g
-             in seq x (Just (ready x, g'))
+             in seq x (Just (x, g'))
   let xs = Seq.fromList $ genericTake w $ unfoldr f g1
-  seq xs (VSeq w $ IndexSeqMap $ (Seq.index xs . fromInteger), g2)
+  seq xs (pure $ VSeq w $ IndexSeqMap $ (Seq.index xs . fromInteger), g2)
 
 -- | Generate a random tuple value.
-randomTuple :: RandomGen g => [Gen g b w i] -> Gen g b w i
+randomTuple :: (Backend sym, RandomGen g) => [Gen g sym] -> Gen g sym
 randomTuple gens sz = go [] gens
   where
-  go els [] g = (VTuple (reverse els), g)
+  go els [] g = (pure $ VTuple (reverse els), g)
   go els (mkElem : more) g =
     let (v, g1) = mkElem sz g
-    in seq v (go (ready v : els) more g1)
+    in seq v (go (v : els) more g1)
 
 -- | Generate a random record value.
-randomRecord :: RandomGen g => Map Ident (Gen g b w i) -> Gen g b w i
+randomRecord :: (Backend sym, RandomGen g) => Map Ident (Gen g sym) -> Gen g sym
 randomRecord gens sz g0 =
-  let (g', m) = Map.mapAccum mk g0 gens in (VRecord m, g')
+  let (g', m) = Map.mapAccum mk g0 gens in (pure $ VRecord m, g')
   where
     mk g gen =
       let (v, g') = gen sz g
-      in seq v (g', ready v)
+      in seq v (g', v)
+
+
+-- Random Values ---------------------------------------------------------------
+
+-- | Produce a random value with the given seed. If we do not support
+-- making values of the given type, return zero of that type.
+-- TODO: do better than returning zero
```

Comment:
Good question.  I'm not sure, but maybe functions and infinite sequences?  It's a little hard to tell.  This TODO predates the refactor, but it might be worth looking into again.

Actually, much of the above could probably be simplified by using `Cryptol.Eval.Type.TValue` instead of directly examining type syntax.

Reply:
Looks like functions are the main type where that case applies.  I'm going to make a separate issue to track this TODO (#695)